includes/wp-admin/class-admin.php

```diff
@@ -593,23 +598,207 @@ public static function user_bulk_options( $actions ) {
 	public static function handle_bulk_request( $send_back, $action, $users ) {
 		if (
 			'remove_activitypub_cap' !== $action &&
-			'add_activitypub_cap' !== $action
+			'add_activitypub_cap' !== $action &&
+			'remove_activitypub_cap_confirmed' !== $action
 		) {
 			return $send_back;
 		}
 
-		foreach ( $users as $user_id ) {
-			$user = new \WP_User( $user_id );
-			if ( 'add_activitypub_cap' === $action ) {
+		// Handle adding ActivityPub capability.
+		if ( 'add_activitypub_cap' === $action ) {
+			foreach ( $users as $user_id ) {
+				$user = new \WP_User( $user_id );
 				$user->add_cap( 'activitypub' );
-			} elseif ( 'remove_activitypub_cap' === $action ) {
+			}
+			return $send_back;
+		}
+
+		// Handle removing ActivityPub capability - remove capabilities immediately, then redirect to fediverse deletion confirmation.
+		if ( 'remove_activitypub_cap' === $action ) {
+			$removed_count = 0;
+
+			// Remove capabilities immediately.
+			foreach ( $users as $user_id ) {
+				$user = new \WP_User( $user_id );
+
+				// Check if user has ActivityPub capability.
+				if ( ! $user->has_cap( 'activitypub' ) ) {
+					continue;
+				}
+
+				// Remove the capability.
 				$user->remove_cap( 'activitypub' );
+
+				// Force cache refresh for user capabilities.
+				\wp_cache_delete( $user_id, 'users' );
+				\wp_cache_delete( $user_id, 'user_meta' );
+
+				++$removed_count;
+			}
+
+			// Build the query args with proper array handling for fediverse deletion confirmation.
+			$query_args = array(
+				'activitypub_confirm_removal' => '1',
```

Comment:
```suggestion
				'action' => 'activitypub_confirm_removal',
```
With this change, could we move `handle_bulk_capability_removal_page()` to the `admin_action_activitypub_confirm_removal` action (instead of `current_screen`)?

That way `handle_bulk_capability_removal_page()` wouldn't have to check for `is_admin()` or whether `activitypub_confirm_removal` is set (and it doesn't execute on every admin page load).